flowauth/backend/flowauth/roles.py

```diff
@@ -0,0 +1,162 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+from flask import Blueprint, jsonify, request, current_app
+import datetime
+
+from flask_login import current_user, login_required
+from flask_principal import Permission, RoleNeed
+
+from .models import Role, Scope, Server, User, db
+
+blueprint = Blueprint(__name__.split(".").pop(), __name__)
+
+admin_permission = Permission(RoleNeed("admin"))
+
+
+def role_to_dict(role):
+    return {
+        "id": role.id,
+        "name": role.name,
+        "scopes": sorted([scope.id for scope in role.scopes]),
+        "latest_token_expiry": role.latest_token_expiry.strftime(
+            "%Y-%m-%dT%H:%M:%S.%fZ"
+        ),
+        "longest_token_life_minutes": role.longest_token_life_minutes,
+        "server": role.server_id,
+        "users": sorted([user.id for user in role.users]),
+    }
+
+
+def roles_to_json(roles):
+    return jsonify(
+        sorted([role_to_dict(role) for role in roles], key=lambda x: x["id"])
+    )
+
+
+@blueprint.route("/", methods=["GET"])
+@login_required
+@admin_permission.require(http_exception=401)
+def list_roles():
+    return jsonify([role_to_dict(role) for role in Role.query.all()])
+
+
+@blueprint.route("/user/<user_id>")
+@login_required
+@admin_permission.require(http_exception=401)
+def list_user_roles(user_id):
+    user = User.query.filter(User.id == user_id).first_or_404()
+    return jsonify([role_to_dict(role) for role in user.roles])
+
+
+@blueprint.route("/<role_id>", methods=["GET"])
+@login_required
+@admin_permission.require(http_exception=401)
+def get_role(role_id):
+    role = Role.query.filter(Role.id == role_id).first()  # First or error?
+    return jsonify(role_to_dict(role))
+
+
+@blueprint.route("/", methods=["POST"])
+@login_required
+@admin_permission.require(http_exception=401)
+def add_role():
+    json = request.get_json()
+    json["latest_token_expiry"] = datetime.datetime.strptime(
+        json["latest_token_expiry"], "%Y-%m-%dT%H:%M:%S.%fZ"
+    )
+    server = Server.query.filter(Server.id == json["server_id"]).first()
+    role_scopes = Scope.query.filter(Scope.id.in_(json["scopes"])).all()
+    try:
+        role_users = User.query.filter(User.id.in_(json["users"])).all()
+    except KeyError:
+        role_users = []
+    new_role = Role(
+        name=json["name"],
+        scopes=role_scopes,
+        users=role_users,
+        server=server,
+        latest_token_expiry=json["latest_token_expiry"],
+        longest_token_life_minutes=json["longest_token_life_minutes"],
```

Comment:
We need to enforce here that the latest expiry and longest lifetime for the role don't exceed those of the server. Previously we had frontend restrictions on this as well (i.e. dates after the server's latest expiry were greyed out in "Create group", and similar for max lifetime), but this at least needs to be enforced in the backend. Previous code that checked this for groups is here: https://github.com/Flowminder/FlowKit/blob/master/flowauth/backend/flowauth/groups.py#L214-L219.

Reply:
In branch + test now